submariner/cable-driver-policy.md

```diff
@@ -0,0 +1,327 @@
+    metav1.TypeMeta `json:",inline"`
+
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec ClusterConnectionPolicySpec `json:"spec"`
+}
+
+type ClusterConnectionPolicySpec struct {
+    // LeftClusterSelector identifies the Cluster resources representing the clusters on one end of a connection. An empty
+    // selector indicates wildcard, ie matches any cluster.
+    LeftClusterSelector metav1.LabelSelector `json:"leftClusterSelector,omitempty"`
+
+    // RightClusterSelector identifies the Cluster resources representing the clusters on the other end of a connection.
+    // An empty selector indicates wildcard, ie matches any cluster.
+    RightClusterSelector metav1.LabelSelector `json:"rightClusterSelector,omitempty"`
+
+    // Name of the cable driver implementation to use for this connection.
+    CableDriver string `json:"CableDriver"`
+
+    // +optional
+    // NoConnectOnConflict defines the intention to fallback to no connection if there's a policy conflict
+    // If defined - should be part of the default policy
+    NoConnectOnConflict bool `json:"NoConnectOnConflict"`
+}
+```
+
+The `ClusterConnectionPolicy` CR defines the criteria for selecting which cable driver to use to connect cluster pairs (or
+subsets). To resolve the connectivity policy for a remote cluster's Endpoint, Submariner would:
+
+- Search for a policy whose left and right label selectors match the corresponding Cluster resources of the local and remote
+  Endpoints.
+
+> **_NOTE:_** In the case of multiple labels being specified, the selector will need to match all the labels.
+> Inequality-based requirements will be supported. For e.g. labels such as `env: !production` will match any clusters
+> that aren't labelled with `env: production`.
+
+If any matching policy is found, the clusters are connected using the cable driver specified in the cable driver policy.
+
+If no matching policy is found, the `default` cable driver policy is used. The `default` policy can be auto-created at broker
+deploy time.
+
+If more than one matching policy is found, the default policy will consulted to establish if `NoConnectOnConflict` is set to true.
+In the case where `NoConnectOnConflict` is true the clusters are **not connected** and the connection status between a pair of clusters
+is marked as `conflict`. In the case where `NoConnectOnConflict` is false, the clusters are connected but the connection status between
+a pair of clusters is marked as `conflict`. If `NoConnectOnConflict` is undefined, it's assumed to be set to True.
+
+The initial `ClusterConnectionPolicy` resources would be created and maintained on the broker cluster and synced to each cluster in the
+ClusterSet.
+
+Submariners `EndpointSpec` needs to be modified as `Backend` and `BackendConfig` only reflect a single cable driver. An alternative parameter
+`Backends` is proposed. This parameter represents a slice of supported backends and their configurations on an Endpoint. This information
```

Comment:
The `BackendConfig` has been overloaded to store NAT discovery and other gateway config in addition to cable driver config.